File: dataplane/internal/core/catalog.go
package core

import "slices"

// MicroUSD is money in millionths of a US dollar.
//
// Integer, never float. Per-token prices are small enough that float64 rounding
// shows up in monthly invoices, and a budget system that disagrees with the bill
// is worse than no budget system.
type MicroUSD int64

// DeploymentID names one concrete, reachable model endpoint.
type DeploymentID string

// RoutingKey identifies a servable target as (base model, adapter).
//
// This is the key the routing table is indexed by, from the first commit, even
// though fine-tuning does not land until much later. The reason is economic: a
// per-tenant fine-tune only pencils out if one vLLM base-model deployment serves
// many LoRA adapters. If the routing key were a bare model name, adding adapters
// later would be a migration of the catalog, the snapshot, and every consumer.
//
// AdapterID is empty for a plain base model. The struct is comparable, so it can
// be a map key directly.
type RoutingKey struct {
	BaseModel string
	AdapterID string
}

// IsAdapter reports whether the key names a fine-tuned adapter.
func (k RoutingKey) IsAdapter() bool { return k.AdapterID != "" }

func (k RoutingKey) String() string {
	if k.AdapterID == "" {
		return k.BaseModel
	}
	return k.BaseModel + "+" + k.AdapterID
}

// Capability is a feature a caller may require of a deployment: streaming, tool
// calling, vision, a context-window class. The router filters candidates by it.
type Capability string

// The capabilities a deployment may declare and a caller may require.
const (
	CapabilityStreaming   Capability = "streaming"
	CapabilityToolCalling Capability = "tool_calling"
	CapabilityVision      Capability = "vision"
	CapabilityEmbeddings  Capability = "embeddings"
)

// Cost is what a deployment charges, per thousand tokens of each class.
//
// The classes exist because providers do not bill all input tokens the same:
// cached input is an order of magnitude cheaper, and writing to a cache is
// dearer than a plain read. A single input rate bills a request that was 90%
// cache reads as though it were not, and the provider reported the truth
// exactly once — in a response nobody kept.
type Cost struct {
	InputPer1K  MicroUSD
	OutputPer1K MicroUSD
	// CachedInputPer1K applies to input served from a provider's prompt cache.
	// Zero means unconfigured and falls back to InputPer1K, which over-bills
	// slightly rather than under-billing — the safe direction for a number a
	// customer disputes.
	CachedInputPer1K MicroUSD
	// CacheWritePer1K applies to input written into a provider's cache, which
	// costs more than a plain read.
	CacheWritePer1K MicroUSD
}

// For returns what a call costs at this price, given what it consumed.
//
// Prices are per thousand tokens, so this divides. Integer division truncates,
// which under-bills by less than a millionth of a dollar per request and never
// over-bills — the right direction for a rounding error a customer can see.
func (c Cost) For(usage TokenUsage) MicroUSD {
	cached := c.CachedInputPer1K
	if cached == 0 {
		// Unconfigured, not free. Billing cached tokens at zero because nobody
		// set a rate would silently give away the majority of a cache-heavy
		// workload.
		cached = c.InputPer1K
	}
	write := c.CacheWritePer1K
	if write == 0 {
		write = c.InputPer1K
	}

	total := MicroUSD(usage.Input)*c.InputPer1K +
		MicroUSD(usage.CachedInput)*cached +
		MicroUSD(usage.CacheWrite)*write +
		MicroUSD(usage.Output)*c.OutputPer1K
	return total / 1000
}

// Deployment is one reachable endpoint that can serve a RoutingKey.
//
// A model with three providers behind it is three Deployments sharing one
// RoutingKey. Selection picks among them; execution walks them in order.
type Deployment struct {
	ID       DeploymentID
	Key      RoutingKey
	Provider string // name of the ProviderPort implementation that speaks to it
	Endpoint string
	Region   string

	TrustTier TrustTier

	// CredentialRef points at a secret; it is never the secret. Provider
	// credentials must not enter a snapshot, because a snapshot is replicated to
	// every worker, cached, and versioned. Workers resolve the reference through
	// the control plane's SecretsPort and cache the result with a TTL.
	CredentialRef string

	// Weight is the share of traffic this deployment may take, 0-100. It is
	// uint32 to match the wire format: a narrower domain type would truncate an
	// out-of-range value into a plausible one instead of failing validation.
	// Zero means
	// registered but not serving: the state a new fine-tuned adapter sits in
	// while it takes shadow traffic, before the canary steps begin.
	Weight uint32

	// ShadowPercent is the share of the base model's traffic mirrored here,
	// 0-100. A separate dimension from Weight: an adapter in shadow serves
	// nobody while seeing real requests, and its response is discarded.
	//
	// Sampled rather than all-or-nothing because mirroring doubles inference
	// spend for whatever fraction it covers, and a shadow that costs as much
	// as production is one an operator turns off.
	ShadowPercent uint32

	Cost         Cost
	Capabilities []Capability
}

// Supports reports whether the deployment offers every required capability.
func (d *Deployment) Supports(required ...Capability) bool {
	for _, want := range required {
		if !slices.Contains(d.Capabilities, want) {
			return false
		}
	}
	return true
}

// Serving reports whether the deployment may take live traffic.
func (d *Deployment) Serving() bool { return d.Weight > 0 }

// Shadowing reports whether this deployment should receive mirrored traffic.
//
// An adapter only: shadowing a base model would mirror a request to the very
// deployment that served it.
func (d *Deployment) Shadowing() bool {
	return d.ShadowPercent > 0 && d.Key.IsAdapter()
}

// ModelAlias decouples client code from concrete model IDs: callers ask for
// "fast" or "reasoning" and the snapshot decides what that means today.
// Targets are in preference order.
type ModelAlias struct {
	Name    string
	Targets []RoutingKey
}

File: controlplane/src/model_gateway_control/domain/catalog.py
"""Model catalog: what can be called, and where it is served."""

from __future__ import annotations

from dataclasses import dataclass, field
from enum import IntEnum, StrEnum

from model_gateway_control.errors import InvalidRequestError


class TrustTier(IntEnum):
    """Where a request is allowed to be sent.

    Ordered, and higher means more trusted, because admission computes a
    *minimum* acceptable tier and the router filters candidates against it.
    ``UNSET`` exists so a deployment cannot silently default to the most
    permissive tier.
    """

    UNSET = 0
    EXTERNAL = 1
    PRIVATE_CLOUD = 2
    INTERNAL = 3


class Capability(StrEnum):
    """A feature a caller may require of a deployment."""

    STREAMING = "streaming"
    TOOL_CALLING = "tool_calling"
    VISION = "vision"
    EMBEDDINGS = "embeddings"


@dataclass(frozen=True, slots=True)
class RoutingKey:
    """Identifies a servable target as (base model, adapter).

    ``adapter_id`` is empty for a plain base model. This is the routing table's
    key from the first release rather than a bare model name, because
    multi-LoRA serving means one base deployment serves many adapters — and
    retrofitting the key later is a migration of the catalog, the snapshot and
    every consumer.

    Frozen so it can be a dictionary key and cannot be mutated after a builder
    has indexed by it.
    """

    base_model: str
    adapter_id: str = ""

    def __post_init__(self) -> None:
        if not self.base_model:
            raise InvalidRequestError("a routing key needs a base model")

    def __str__(self) -> str:
        return f"{self.base_model}+{self.adapter_id}" if self.adapter_id else self.base_model


@dataclass(frozen=True, slots=True)
class Cost:
    """Price per thousand tokens, in millionths of a US dollar.

    Integer, never float: per-token prices are small enough that float rounding
    reaches the invoice.
    """

    input_per_1k_micro_usd: int = 0
    output_per_1k_micro_usd: int = 0

    def __post_init__(self) -> None:
        if self.input_per_1k_micro_usd < 0 or self.output_per_1k_micro_usd < 0:
            raise InvalidRequestError("a price cannot be negative")


@dataclass(frozen=True, slots=True, kw_only=True)
class Deployment:
    """One reachable endpoint that can serve a routing key.

    Keyword-only because a positional constructor with this many same-typed
    string fields is unreadable and fragile to reordering.
    """

    id: str
    key: RoutingKey
    provider: str
    endpoint: str
    region: str = ""
    trust_tier: TrustTier = TrustTier.UNSET
    # A reference to a secret, never the secret. Credentials must not enter a
    # snapshot: snapshots are replicated to every worker, cached and versioned.
    credential_ref: str = ""
    # Share of traffic, 0-100. Zero means registered but not serving — where a
    # fine-tuned adapter sits while it takes shadow traffic.
    weight: int = 0
    cost: Cost = field(default_factory=Cost)
    capabilities: tuple[Capability, ...] = ()

    def __post_init__(self) -> None:
        if not self.id:
            raise InvalidRequestError("a deployment needs an id")
        if not self.provider:
            raise InvalidRequestError(f"deployment {self.id!r} names no provider")
        if self.trust_tier is TrustTier.UNSET:
            raise InvalidRequestError(f"deployment {self.id!r} has an unset trust tier")
        if not 0 <= self.weight <= 100:
            raise InvalidRequestError(
                f"deployment {self.id!r} weight {self.weight} is out of range"
            )


@dataclass(frozen=True, slots=True)
class ModelAlias:
    """Decouples client code from concrete model ids.

    Callers ask for ``fast`` or ``reasoning``; the snapshot decides what that
    means today. Targets are in preference order.
    """

    name: str
    targets: tuple[RoutingKey, ...]

    def __post_init__(self) -> None:
        if not self.name:
            raise InvalidRequestError("an alias needs a name")
        if not self.targets:
            raise InvalidRequestError(f"alias {self.name!r} has no targets")
